bson/src/main/org/bson/codecs/StringCodec.java

```diff
@@ -16,27 +16,74 @@
 
 package org.bson.codecs;
 
+import org.bson.BsonInvalidOperationException;
 import org.bson.BsonReader;
 import org.bson.BsonType;
 import org.bson.BsonWriter;
+import org.bson.codecs.configuration.CodecConfigurationException;
+import org.bson.types.ObjectId;
+
+import static org.bson.assertions.Assertions.isTrueArgument;
 
 /**
  * Encodes and decodes {@code String} objects.
  *
  * @since 3.0
  */
-public class StringCodec implements Codec<String> {
+public class StringCodec implements Codec<String>, RepresentationConfigurable<String> {
+    private BsonType representation;
+
+    /**
+     * Constructs a StringCodec with a String representation.
+     */
+    public StringCodec() {
+        representation = BsonType.STRING;
+    }
+
+    private StringCodec(final BsonType representation) {
+        this.representation = representation;
+    }
+
+    @Override
+    public BsonType getRepresentation() {
+        return representation;
+    }
+
+    @Override
+    public Codec<String> withRepresentation(final BsonType representation) {
+        isTrueArgument("representation = ObjectId or String",
+                representation == BsonType.OBJECT_ID || representation == BsonType.STRING);
+        return new StringCodec(representation);
+    }
+
+
     @Override
     public void encode(final BsonWriter writer, final String value, final EncoderContext encoderContext) {
-        writer.writeString(value);
+        switch (representation) {
+            case STRING:
+                writer.writeString(value);
+                break;
+            case OBJECT_ID:
+                writer.writeObjectId(new ObjectId(value));
+                break;
+            default:
+                throw new BsonInvalidOperationException("Cannot encode a String to a " + representation);
```

Comment:
This line is pretty much unreachable now (because the only representations `StringCodec` allows to be created are  `STRING` and `OBJECT_ID`. Is this kind of defensive programming good or unnecessary?

Reply:
I think this is good defensive programming, since a change in a different method could break the invariants here, so better to be defensive.  And it would be bad to silently do nothing in this method if somehow the type was unsupported.